src/e2e_tests.rs

```diff
@@ -0,0 +1,649 @@
+use super::identity::{Identity, Signer};
+use super::processor;
+use super::proxy::ProxyWorker;
+use super::redis;
+use super::storage::{ProxyStorage, RedisStorage, Storage};
+use super::twin::{Twin, TwinDB};
+use super::types::Message;
+use anyhow::{Context, Result};
+use bb8_redis::redis::AsyncCommands;
+use sp_core::crypto::Pair;
+use sp_core::{ed25519::Pair as EdPair, sr25519::Pair as SrPair};
+use std::collections::HashMap;
+use std::panic;
+use std::process::Command;
+use std::sync::Once;
+use tokio::time::{sleep, Duration};
+
+#[derive(Default, Clone)]
+struct InMemoryDB {
+    pub twins: HashMap<u32, Twin>,
+}
+
+impl InMemoryDB {
+    fn add(&mut self, twin: Twin) {
+        self.twins.insert(twin.id, twin);
+    }
+}
+
+#[async_trait::async_trait]
+impl TwinDB for InMemoryDB {
+    async fn get_twin(&self, twin_id: u32) -> anyhow::Result<Option<Twin>> {
+        Ok(self.twins.get(&twin_id).map(|t| t.clone()))
+    }
+
+    async fn get_twin_with_account(
+        &self,
+        _account_id: sp_runtime::AccountId32,
+    ) -> anyhow::Result<u32> {
+        unimplemented!()
+    }
+}
+
+fn new_message(
+    command: &str,
+    expiration: u64,
+    retry: usize,
+    data: &str,
+    destination: Vec<u32>,
+) -> Message {
+    let ret_queue = format!("{}", uuid::Uuid::new_v4());
+    let epoch = std::time::SystemTime::now()
+        .duration_since(std::time::SystemTime::UNIX_EPOCH)
+        .unwrap()
+        .as_secs();
+    Message {
+        command: String::from(command),
+        expiration: expiration,
+        retry: retry,
+        data: data.to_string(),
+        destination: destination,
+        reply: ret_queue,
+        timestamp: epoch,
+        ..Message::default()
+    }
+}
+
+async fn send_all(messages: Vec<Message>, local_redis: &str) -> Result<(usize, Vec<String>)> {
+    let pool = redis::pool(local_redis)
+        .await
+        .context("unable to create redis connection")?;
+    let mut conn = pool.get().await.context("unable to get redis connection")?;
+    let queue = "msgbus.system.local";
+    let mut responses_expected = 0;
+    let mut return_queues = Vec::new();
+    for msg in messages {
+        let _ = conn.lpush(&queue, &msg).await?;
+        responses_expected += msg.destination.len();
+        return_queues.push(msg.reply);
+    }
+    Ok((responses_expected, return_queues))
+}
+
+async fn wait_for_responses(
+    responses_expected: usize,
+    return_queues: Vec<String>,
+    timeout: usize,
+    local_redis: &str,
+) -> Result<(Vec<Message>, usize, usize, usize)> {
+    let pool = redis::pool(local_redis)
+        .await
+        .context("unable to create redis connection")?;
+    let mut conn = pool.get().await.context("unable to get redis connection")?;
+    let mut responses = Vec::new();
+    let mut err_count = 0;
+    let mut success_count = 0;
+    for _ in 0..responses_expected {
+        let result: Option<(String, String)> = conn
+            .blpop(&return_queues, timeout)
```

Comment:
`timeout` calculation here is wrong. Let's say you called this method with a timeout `60` seconds. So you expect this method to wait total MAX time of 60 seconds. Now let's assume you are waiting for 10 messages to come back. Let's also assume that each messages takes 5 seconds to come back.

 so you wait 5 seconds, then u get first response, then u wait another 60 seconds. then u get another one after only 5 seconds, then u wait another 60 seconds, and so on
so the total wait will be (5*10) + 60 which is 110 seconds.

I know this is not that important in a test scenario,Hence it's okay to ignore it. But i just wanted you to be aware of the problem for the future.

Reply:
yes, I understand. but this was not intended to be on par with the message TTL. it was only a cheap-to-implement way to make sure the client/test won't wait forever for the responses and that it will timeout eventually if it doesn't receive a response for a while. I even wrote a comment in the caller function explaining this.
https://github.com/threefoldtech/rmb-rs/blob/d0333efd2897a99cd9d43313fdbc0463be5716b8/src/e2e_tests.rs#L143

resolving it as no need to modify as you mentioned.